### PR TITLE
Add admin API for logging in as a user

### LINKED ISSUE #6054: Admin API to return an access token for any user of the server
**Description:**

It would be useful for the server administrator to be able to grab an access token on behalf of any user of their server. As discussed with @ara4n it would be good to extend the 'whois' admin API to offer this functionality

This could be useful for admins of the server to perform actions on behalf of another user (e.g. a dedicated 'abuse' user or bots etc.) and could also be useful for (benign) account puppeting, e.g. provisioning of integrations etc.

### PR DESCRIPTION
**Pull request by erikjohnston**
~~Based on #8616, #8614~~. Fixes #6054.

I think reviewing this commit by commit makes the most sense. The PR is a bit large, I'm hesitant to split it up but can do if people struggle.

### CODE REVIEW

File: synapse/rest/admin/users.py [hunk 1]
```diff
@@ -826,3 +829,42 @@ async def on_GET(
...
+        if not self.hs.is_mine_id(user_id):
+            raise SynapseError(400, "Only local users can be logged in as")
+
+        body = parse_json_object_from_request(request)
+
+        valid_until_ms = body.get("valid_until_ms")
+        if valid_until_ms and not isinstance(valid_until_ms, int):
+            raise SynapseError(400, "'valid_until_ms' parameter must be an int")
+
+        if auth_user.to_string() == user_id:
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/rest/admin/users.py:860]: User IDs are case sensitive, we fudge it a bit at log in but that's about it. Plus I don't think its the end of the world to puppet it yourself, its just a weird thing to try and do.

File: synapse/handlers/message.py [hunk 2]
```diff
@@ -619,7 +619,13 @@ async def assert_accepted_privacy_policy(self, requester: Requester) -> None:
         if requester.app_service is not None:
             return
 
-        user_id = requester.user.to_string()
+        user_id = requester.authenticated_entity
+        if not user_id.startswith("@"):
+            # The authenticated entity might not be a user, e.g. if it's the
+            # server puppetting the user.
+            return
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/message.py:626]: As you saw we now use the local server name for when the server is messing around with users.

File: tests/rest/admin/test_user.py [hunk 3]
```diff
@@ -1521,3 +1521,244 @@ def _check_fields(self, content):
...
+        # Create a room as the admin user. This will bump the monthly active users to 1.
+        room_id = self.helper.create_room_as(self.admin_user, tok=self.admin_user_tok)
+
+        # Trying to join as the other user should fail.
+        self.helper.join(
+            room_id, user=self.other_user, tok=self.other_user_tok, expect_code=403
+        )
+
+        # Logging in as the other user and joining a room should work, even
+        # though they should be denied.
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply tests/rest/admin/test_user.py:1762]: I've rejigged the wording. The MAU limit would stop the real user from joining (or doing other actions) as the server has reached the MAU limit, but the admin user is still allowed to puppet them. The MAU isn't bumped to 2 as we base those off the authenticated entity, which is always the admin user in this case.

File: synapse/handlers/profile.py [hunk 4]
```diff
@@ -204,7 +204,9 @@ async def set_displayname(
         # the join event to update the displayname in the rooms.
         # This must be done by the target user himself.
         if by_admin:
-            requester = create_requester(target_user)
+            requester = create_requester(
+                target_user, authenticated_entity=requester.authenticated_entity,
+            )
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/profile.py:209]: It's a bit of an edge case tbh. The original request has kinda the "right" requester, we could fudge it but then it is a bit less clear whether the admin user is using a puppeted token vs authenticating as themselves :shrug:

File: synapse/rest/admin/users.py [hunk 5]
```diff
@@ -826,3 +829,42 @@ async def on_GET(
...
+
+    PATTERNS = admin_patterns("/users/(?P<user_id>[^/]*)/login$")
+
+    def __init__(self, hs: "HomeServer"):
+        self.hs = hs
+        self.store = hs.get_datastore()
+        self.auth = hs.get_auth()
+        self.auth_handler = hs.get_auth_handler()
+
+    async def on_PUT(self, request, user_id):
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/rest/admin/users.py:846]: Yes, and its not idempotent so `PUT` is just wrong.

File: synapse/handlers/register.py [hunk 6]
```diff
@@ -362,7 +366,9 @@ async def _create_and_join_rooms(self, user_id: str):
                     # created it, then ensure the first user joins it.
                     if requires_join:
                         await room_member_handler.update_membership(
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/handlers/register.py:368]: No, and we catch all exception anyway so may as well remove it.

File: synapse/module_api/__init__.py [hunk 7]
```diff
@@ -336,7 +337,9 @@ async def create_and_send_event_into_room(self, event_dict: JsonDict) -> EventBa
             SynapseError if the event was not allowed.
         """
         # Create a requester object
-        requester = create_requester(event_dict["sender"])
+        requester = create_requester(
+            event_dict["sender"], authenticated_entity=self._server_name
+        )
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/module_api/__init__.py:342]: Yes :+1:, in the sense that they aren't subject to rate limits or MAU limits, which I think is what you'd expect.

**Comment by Twi1ightSparkle** [comment]
Will this also allow logging in as `@server:homeserver.tld`? I understand that this is a bit of a special user?

**Comment by anoadragon453** [comment]
> Will this also allow logging in as @server:homeserver.tld? I understand that this is a bit of a special user?

I believe this is the `auto_join_mxid_localpart` user ([docs](https://github.com/matrix-org/synapse/blob/26b46796ea7d092c2fc3f47be7923c3a2cf634d3/docs/sample_config.yaml#L1341)) that you're referring to. Yes, this should allow retrieving an access token for this user if I understand things correctly.

File: synapse/rest/admin/users.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (clokep) [reply synapse/rest/admin/users.py:870]: Well the case sensitivity is only there for backwards compatibility, I think? But yeah I get your point of it not *really* being an error, just weird.

**Review by clokep — APPROVED** [review]
I think this looks fine. One thought though -- is there a way to get a list of your access tokens that refer to other accounts? This seems like it would be a useful thing to look up after the fact? (E.g. if you wanted to have an admin API to logout of other access tokens easily.)

**Comment by clokep** [comment]
This will need `develop` merged as we discussed! 😄

**Comment by erikjohnston** [comment]
> I think this looks fine. One thought though -- is there a way to get a list of your access tokens that refer to other accounts? This seems like it would be a useful thing to look up after the fact? (E.g. if you wanted to have an admin API to logout of other access tokens easily.)

Yeah, probably, though I'm inclined to leave that for now.

**Pull request merged**